Expose rpc.Future.set_result() as a Python API

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37638 Remove unnecessary pickle and unpickle invocation in PyRRef __setstate__/__getstate__ methods
* **#37607 Expose rpc.Future.set_result() as a Python API**
* #37311 Let rpc.Future expose an add_done_callback API

Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

Commit:
Expose rpc.Future.set_result() as a Python API

[ghstack-poisoned]

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"




[ghstack-poisoned]

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"


Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

[ghstack-poisoned]

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"


Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -405,6 +406,42 @@ If the future completes with an error, an exception is thrown.
+              applications should never call this method on a ``Future``
+              returned by :meth:`~torch.distributed.rpc.rpc_async`, because the
+              RPC system will also set result for that ``Future`` when the
+              response is received. This method should only be called on
```
Do we know what error will happen if the user accidentally does this? I feel like we should add a check if possible to ensure that this doesn't happen - is there a way to distinguish between user's future and RPC system future easily?

If not, then for now we can also consider checking `fut.isCompleted()` in set_result, and raise an error to the user indicating they may be trying to use it incorrectly. Although, this wouldn't catch the case where the user calls set_result, and after that, the RPC system completes the future

mrshenli replied:
Let me add a test to check the error message. Current code in `future.h` already checks `completed_` in `markCompleted()`. Will make sure the error message is clear.

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -310,6 +310,25 @@ def callback(fut):
+        fut2 = rpc.rpc_async(to, torch.add, args=(fut1.wait(), z))
+        fut2.add_done_callback(lambda fut2 : out.set_result(fut2.wait()))
+
+    fut1 = rpc.rpc_async(to, torch.add, args=(x, y))
```
Do we want to test calling UDFs that expect `rpc.Future` arguments?

mrshenli replied:
Good point, this is currently not a targeted use case. Let me customize `Future`'s pickle methods to through the correct error.

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -405,6 +406,42 @@ If the future completes with an error, an exception is thrown.
                   >>> fut.add_done_callback(callback)
+          )")
+      .def(
+          "set_result",
```
Is it an explicit goal to mimic `concurrent.Futures` API? If so, should we have `get()` as well?

mrshenli replied:
Yes, according to the discussion in #36071, Python users might be familiar with this set of API. Let me add that. You mean `Future.result()` instead of `Future.get()`?

mrshenli replied:
Hmm, this will then be duplicated with the existing `wait` API. Let's postpone that until we discussed this with the team.

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"


Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

[ghstack-poisoned]

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"


Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

[ghstack-poisoned]

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"


Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

[ghstack-poisoned]

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"


Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

[ghstack-poisoned]

Commit:
Update on "Expose rpc.Future.set_result() as a Python API"


Differential Revision: [D21336330](https://our.internmc.facebook.com/intern/diff/D21336330)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2432,6 +2455,75 @@ def test_callback_in_rpc(self):
         )
         self.assertEqual(ret, torch.ones(2, 2) + 1 + 2)
 
+    @dist_init
+    def test_future(self):
+        f = rpc.Future()
+        f.set_result(torch.ones(2, 2))
```
We should probably add some torchscript tests for this once the Futures are merged.

mrshenli commented:
Superseded by #39008
